sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala

```diff
@@ -100,21 +113,25 @@ object GpuSemaphore {
 
 private final class GpuSemaphore(tasksPerGpu: Int) extends Logging with Arm {
   private val semaphore = new Semaphore(tasksPerGpu)
+
   // Map to track which tasks have acquired the semaphore.
-  private val activeTasks = new ConcurrentHashMap[Long, MutableInt]
+  case class TaskInfo(count: MutableInt, thread: Thread)
```

Comment:
I hadn't taken the python code into account at all. In this scenario, are both threads allocating memory? What you are describing doesn't seem to fit with the semaphore. I need to read that code a bit to understand it.

Reply:
> My only nits are that we assume that an OOM is fatal

we can signify this via a more fine-grained exception types. Also will help to stop misappropriating `java.lang.OutOfMemoryError` that specifically relates to the JVM memory https://docs.oracle.com/javase/7/docs/api/java/lang/OutOfMemoryError.html 

We should have our own class hierarchy for GPU Memory exceptions, 
ai.rapids.cudf.GpuOutOfMemory -> FatalGpuOutOfMemory

Reply:
@revans2 and I discussed locally some of this. I think the main worry is: 

1. Thread 1 gets the semaphore, and therefore is registered against the `taskAttemptId` -> `TaskInfo` map.
2. Thread 1 releases the semaphore. The task is not complete yet, so the entry in (1) is still there.
3. Thread 2 gets the semaphore (for the same task), and so the reference count is increased in `TaskInfo` but the new thread information is not captured. In order to capture this, we'd want the second thread to add itself to a `Set[Thread]`, so we keep the info.
4. OOM happens, we only see the info for the first thread, missing the potential issue in thread 2.

I'd say that as a corner case it seems like something we can improve tracking a bit, but I also believe that a stack trace may put you pretty close to where you need to be. Thread 1 may be clue enough that it is this python interaction that is causing the OOM, and then we need to investigate further. The idea with this extra logging is help to narrow down code that is using too much memory.

Reply:
> We should have our own class hierarchy for GPU Memory exceptions

Yes it would be good to not overload `java.lang.OutOfMemoryError`. One use for `FatalGpuOutOfMemory` would be something that tasks can throw after they attempt to handle a non-fatal-oom (point 4 here https://github.com/NVIDIA/spark-rapids/issues/6785).

Reply:
Filed this for the python thread tracking for now: https://github.com/NVIDIA/spark-rapids/issues/6817.

Reply:
Filed this one @gerashegalov for the exception as you suggested https://github.com/rapidsai/cudf/issues/11970. I think we want to add the attempted size (that failed) in order of this exception to make its way to the driver and to be really explicit in the executor logs.